In [1]:
# Import the functions used in this project
from private_pkg.functions import *

# Pretty display for notebooks
%matplotlib inline

# Ignore the warnings
import warnings
warnings.filterwarnings('ignore')

# Load the dataset
train = pd.read_csv("../asset/train.csv")
test = pd.read_csv("../asset/test.csv")
submission = pd.read_csv("../asset/sample_submission.csv")

# Success - Display the first record
print("Train data : ", train.shape)
print("Test  data : ", test.shape)
print("submission  data : ", submission.shape)

# 원본을 유지하기 위해서 카피
df_train = train.copy()
df_test = test.copy()
df_submission = submission.copy()

Train data :  (647054, 7)
Test  data :  (653646, 6)
submission  data :  (95674, 39)


In [5]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.naive_bayes import MultinomialNB
def fitNaiveBayesModel(X, y):
    return MultinomialNB().fit(X, y)

In [7]:
train_fl_li = df_train["FinelineNumber"].unique()
test_fl_li = df_test["FinelineNumber"].unique()
dd_cols = [li if not np.isnan(li) else -1.0 for li in list(set(list(test_fl_li) + list(train_fl_li)))]

In [8]:
X = df_train_model[dd_cols]

In [9]:
y = df_train_model["TripType"]

In [10]:
model_fl = fitNaiveBayesModel(X, y)

In [11]:
y_pred_fl = model_fl.predict(X)

In [15]:
np.trace(confusion_matrix(y, y_pred_fl))/len(df_train_model)

0.5467734180655142

In [28]:
print(classification_report(y, y_pred_fl))

             precision    recall  f1-score   support

          3       0.87      0.95      0.91      3643
          4       0.71      0.26      0.38       346
          5       0.72      0.71      0.72      4593
          6       0.71      0.59      0.64      1277
          7       0.69      0.72      0.71      5752
          8       0.39      0.59      0.47     12161
          9       0.62      0.58      0.60      9464
         12       0.56      0.34      0.42       269
         14       1.00      0.75      0.86         4
         15       0.52      0.60      0.56       978
         18       0.60      0.42      0.50       549
         19       0.83      0.25      0.38       375
         20       0.69      0.46      0.55       637
         21       0.59      0.65      0.62       641
         22       0.73      0.43      0.54       928
         23       0.84      0.12      0.20       139
         24       0.59      0.60      0.59      2609
         25       0.60      0.68      0.64   

In [25]:
X_fl_test = df_test_model[dd_cols]
y_test_pred_fl = model_fl.predict(X_fl_test)

In [26]:
df_submission_2 = makeSubmissionDf(submission, y_test_pred_fl)

0명 진행됨. 아직 95674명 데이터 남음.
5000명 진행됨. 아직 90674명 데이터 남음.
10000명 진행됨. 아직 85674명 데이터 남음.
15000명 진행됨. 아직 80674명 데이터 남음.
20000명 진행됨. 아직 75674명 데이터 남음.
25000명 진행됨. 아직 70674명 데이터 남음.
30000명 진행됨. 아직 65674명 데이터 남음.
35000명 진행됨. 아직 60674명 데이터 남음.
40000명 진행됨. 아직 55674명 데이터 남음.
45000명 진행됨. 아직 50674명 데이터 남음.
50000명 진행됨. 아직 45674명 데이터 남음.
55000명 진행됨. 아직 40674명 데이터 남음.
60000명 진행됨. 아직 35674명 데이터 남음.
65000명 진행됨. 아직 30674명 데이터 남음.
70000명 진행됨. 아직 25674명 데이터 남음.
75000명 진행됨. 아직 20674명 데이터 남음.
80000명 진행됨. 아직 15674명 데이터 남음.
85000명 진행됨. 아직 10674명 데이터 남음.
90000명 진행됨. 아직 5674명 데이터 남음.
95000명 진행됨. 아직 674명 데이터 남음.


In [27]:
saveDataFrameToCsv(df_submission_2, "second_submit")

In [10]:
cols = [uq if not type(uq) is float else "Null" for uq in df_train["DepartmentDescription"].unique()]

In [9]:
X = df[cols]

In [10]:
y = df["TripType"]

In [11]:
model_mult = MultinomialNB().fit(X, y)

In [31]:
y_pred = model_mult.predict(X)

NameError: name 'model_mult' is not defined

In [15]:
print(classification_report(y, y_pred))

             precision    recall  f1-score   support

          3       1.00      0.00      0.00      3643
          4       0.00      0.00      0.00       346
          5       0.56      0.00      0.00      4593
          6       0.00      0.00      0.00      1277
          7       0.00      0.00      0.00      5752
          8       0.15      1.00      0.26     12161
          9       0.00      0.00      0.00      9464
         12       0.00      0.00      0.00       269
         14       0.12      0.25      0.17         4
         15       0.79      0.02      0.05       978
         18       1.00      0.00      0.00       549
         19       0.00      0.00      0.00       375
         20       0.00      0.00      0.00       637
         21       0.80      0.01      0.01       641
         22       0.00      0.00      0.00       928
         23       0.00      0.00      0.00       139
         24       1.00      0.00      0.00      2609
         25       0.80      0.00      0.00   

In [16]:
np.trace(confusion_matrix(y, y_pred))/len(df)

0.1952568095825407

In [8]:
df_1 = pd.read_csv("df_we_wanted_201807251743.csv")

In [11]:
X = df_1[cols]
y = df_1["TripType"]

In [15]:
model_mult_1 = MultinomialNB().fit(X, y)

In [16]:
y_pred_1 = model_mult_1.predict(X)

In [17]:
confusion_matrix(y, y_pred_1)

array([[3163,    0,    9, ...,    0,    0,  403],
       [   0,   13,  278, ...,    0,    0,    7],
       [   0,   96, 3737, ...,    4,    2,   65],
       ...,
       [   0,    3,   28, ...,   33,   30,   73],
       [   0,    2,    9, ...,   58,  202,   35],
       [ 705,    2,   81, ...,    2,    0,  291]])

In [20]:
np.trace(confusion_matrix(y, y_pred_1))/len(df_1)

0.4660513828208291

In [19]:
print(classification_report(y, y_pred_1))

             precision    recall  f1-score   support

          3       0.80      0.87      0.84      3643
          4       0.07      0.04      0.05       346
          5       0.75      0.81      0.78      4593
          6       0.48      0.55      0.51      1277
          7       0.60      0.66      0.63      5752
          8       0.36      0.52      0.42     12161
          9       0.57      0.45      0.50      9464
         12       0.15      0.21      0.18       269
         14       0.05      0.50      0.09         4
         15       0.37      0.49      0.42       978
         18       0.38      0.53      0.44       549
         19       0.39      0.23      0.29       375
         20       0.47      0.73      0.57       637
         21       0.41      0.76      0.53       641
         22       0.45      0.22      0.30       928
         23       0.27      0.20      0.23       139
         24       0.51      0.64      0.57      2609
         25       0.56      0.74      0.64   

In [5]:
df_test = test.copy()

In [6]:
df_test_we_wanted = makeDfWeWanted(df_test, df_train, is_use_negative_scan_count=False)

##### makeDfWeWanted함수는 !! 데이터프레임을 반환하는     함수이므로 변수로 받아주셔야합니다!! 오래 걸리는 작업(맥북 프로 기준 3분)이므로 지금 실수하셨다면 빨리 커널 종료하고 다시 시도해주세요.

0명 진행됨. 아직 95674명 데이터 남음.
5000명 진행됨. 아직 90674명 데이터 남음.
10000명 진행됨. 아직 85674명 데이터 남음.
15000명 진행됨. 아직 80674명 데이터 남음.
20000명 진행됨. 아직 75674명 데이터 남음.
25000명 진행됨. 아직 70674명 데이터 남음.
30000명 진행됨. 아직 65674명 데이터 남음.
35000명 진행됨. 아직 60674명 데이터 남음.
40000명 진행됨. 아직 55674명 데이터 남음.
45000명 진행됨. 아직 50674명 데이터 남음.
50000명 진행됨. 아직 45674명 데이터 남음.
55000명 진행됨. 아직 40674명 데이터 남음.
60000명 진행됨. 아직 35674명 데이터 남음.
65000명 진행됨. 아직 30674명 데이터 남음.
70000명 진행됨. 아직 25674명 데이터 남음.
75000명 진행됨. 아직 20674명 데이터 남음.
80000명 진행됨. 아직 15674명 데이터 남음.
85000명 진행됨. 아직 10674명 데이터 남음.
90000명 진행됨. 아직 5674명 데이터 남음.
95000명 진행됨. 아직 674명 데이터 남음.


In [25]:
X_test = df_test_we_wanted[cols]

In [26]:
y_test_pred = model_mult_1.predict(X_test)

In [27]:
y_test_pred

array([ 38,   8,   8, ..., 999,  32,  25])

In [28]:
saveDataFrameToCsv(df_test_we_wanted, "df_test_we_wanted")

In [29]:
submission

VisitNumber  TripType_3  TripType_4  TripType_5  TripType_6  \
0                1           0           0           0           0   
1                2           0           0           0           0   
2                3           0           0           0           0   
3                4           0           0           0           0   
4                6           0           0           0           0   
5               13           0           0           0           0   
6               14           0           0           0           0   
7               16           0           0           0           0   
8               18           0           0           0           0   
9               21           0           0           0           0   
10              22           0           0           0           0   
11              24           0           0           0           0   
12              27           0           0           0           0   
13              34           0           0           0           0   
14              35           0           0           0           0   
15              36           0           0           0           0   
16              37           0           0           0           0   
17              38           0           0           0           0   
18              39           0           0           0           0   
19              44           0           0           0           0   
20              46           0           0           0           0   
21              48           0           0           0           0   
22              52           0           0           0           0   
23              58           0           0           0           0   
24              59           0           0           0           0   
25              60           0           0           0           0   
26              62           0           0           0           0   
27              64           0           0           0           0   
28              65           0           0           0           0   
29              66           0           0           0           0   
...            ...         ...         ...         ...         ...   
95644       191299           0           0           0           0   
95645       191300           0           0           0           0   
95646       191302           0           0           0           0   
95647       191303           0           0           0           0   
95648       191306           0           0           0           0   
95649       191307           0           0           0           0   
95650       191308           0           0           0           0   
95651       191309           0           0           0           0   
95652       191310           0           0           0           0   
95653       191314           0           0           0           0   
95654       191315           0           0           0           0   
95655       191316           0           0           0           0   
95656       191317           0           0           0           0   
95657       191319           0           0           0           0   
95658       191320           0           0           0           0   
95659       191321           0           0           0           0   
95660       191323           0           0           0           0   
95661       191325           0           0           0           0   
95662       191327           0           0           0           0   
95663       191328           0           0           0           0   
95664       191330           0           0           0           0   
95665       191332           0           0           0           0   
95666       191333           0           0           0           0   
95667       191334           0           0           0           0   
95668       191336           0           0           0           0   
95669       19133

In [30]:
y_test_pred

array([ 38,   8,   8, ..., 999,  32,  25])

In [35]:
submission

VisitNumber  TripType_3  TripType_4  TripType_5  TripType_6  \
0                1           1           0           0           0   
1                2           0           0           0           0   
2                3           0           0           0           0   
3                4           0           0           0           0   
4                6           0           0           0           0   
5               13           0           0           0           0   
6               14           0           0           0           0   
7               16           0           0           0           0   
8               18           0           0           0           0   
9               21           0           0           0           0   
10              22           0           0           0           0   
11              24           0           0           0           0   
12              27           0           0           0           0   
13              34           0           0           0           0   
14              35           0           0           0           0   
15              36           0           0           0           0   
16              37           0           0           0           0   
17              38           0           0           0           0   
18              39           0           0           0           0   
19              44           0           0           0           0   
20              46           0           0           0           0   
21              48           0           0           0           0   
22              52           0           0           0           0   
23              58           0           0           0           0   
24              59           0           0           0           0   
25              60           0           0           0           0   
26              62           0           0           0           0   
27              64           0           0           0           0   
28              65           0           0           0           0   
29              66           0           0           0           0   
...            ...         ...         ...         ...         ...   
95644       191299           0           0           0           0   
95645       191300           0           0           0           0   
95646       191302           0           0           0           0   
95647       191303           0           0           0           0   
95648       191306           0           0           0           0   
95649       191307           0           0           0           0   
95650       191308           0           0           0           0   
95651       191309           0           0           0           0   
95652       191310           0           0           0           0   
95653       191314           0           0           0           0   
95654       191315           0           0           0           0   
95655       191316           0           0           0           0   
95656       191317           0           0           0           0   
95657       191319           0           0           0           0   
95658       191320           0           0           0           0   
95659       191321           0           0           0           0   
95660       191323           0           0           0           0   
95661       191325           0           0           0           0   
95662       191327           0           0           0           0   
95663       191328           0           0           0           0   
95664       191330           0           0           0           0   
95665       191332           0           0           0           0   
95666       191333           0           0           0           0   
95667       191334           0           0           0           0   
95668       191336           0           0           0           0   
95669       19133

In [46]:
def makeSubmissionDf(df, y_pred):
    for i, df_row in enumerate(df.index):
        col_name = "TripType_" + str(y_pred[i])
        df.loc[df_row][col_name] = 1
        if (i % 5000) == 0:    
            print(str(i) + "명 진행됨. 아직 " + str(len(df) - i) + "명 데이터 남음.")
    return df

In [47]:
df_submission_1 = makeSubmissionDf(df_submission, y_test_pred)

0명 진행됨. 아직 95674명 데이터 남음.
5000명 진행됨. 아직 90674명 데이터 남음.
10000명 진행됨. 아직 85674명 데이터 남음.
15000명 진행됨. 아직 80674명 데이터 남음.
20000명 진행됨. 아직 75674명 데이터 남음.
25000명 진행됨. 아직 70674명 데이터 남음.
30000명 진행됨. 아직 65674명 데이터 남음.
35000명 진행됨. 아직 60674명 데이터 남음.
40000명 진행됨. 아직 55674명 데이터 남음.
45000명 진행됨. 아직 50674명 데이터 남음.
50000명 진행됨. 아직 45674명 데이터 남음.
55000명 진행됨. 아직 40674명 데이터 남음.
60000명 진행됨. 아직 35674명 데이터 남음.
65000명 진행됨. 아직 30674명 데이터 남음.
70000명 진행됨. 아직 25674명 데이터 남음.
75000명 진행됨. 아직 20674명 데이터 남음.
80000명 진행됨. 아직 15674명 데이터 남음.
85000명 진행됨. 아직 10674명 데이터 남음.
90000명 진행됨. 아직 5674명 데이터 남음.
95000명 진행됨. 아직 674명 데이터 남음.


In [49]:
saveDataFrameToCsv(df_submission_1, "first_submit")

In [ ]:
train_fl_li = df_train["FinelineNumber"].unique()
test_fl_li = df_test["FinelineNumber"].unique()


In [73]:
train_fl_li = df_train["FinelineNumber"].unique()
test_fl_li = df_test["FinelineNumber"].unique()
train_test_fl_union = [li if not np.isnan(li) else -1.0 for li in list(set(list(test_fl_li) + list(train_fl_li)))]


In [61]:
len(test_fl_li), len(train_fl_li)

(5204, 5196)

In [68]:
len(set(list(test_fl_li) + list(train_fl_li)))

5355

In [52]:
diff_li = []
for fl in test_fl_li:
    if fl not in train_fl_li:
        diff_li.append(fl)

In [55]:
diff_li_train_to_test = []
for fl in train_fl_li:
    if fl not in test_fl_li:
        diff_li_train_to_test.append(fl)

In [57]:
len(diff_li_train_to_test)

151

In [58]:
len(diff_li)

159

In [23]:
df_test_model = makeDfWeWanted_local(df = df_test, df_train = df_train, df_test = df_test, group_by_col_name="FinelineNumber")

##### makeDfWeWanted함수는 !! 데이터프레임을 반환하는     함수이므로 변수로 받아주셔야합니다!! 오래 걸리는 작업(맥북 프로 기준 3분)이므로 지금 실수하셨다면 빨리 커널 종료하고 다시 시도해주세요.

0명 진행됨. 아직 95674명 데이터 남음.
5000명 진행됨. 아직 90674명 데이터 남음.
10000명 진행됨. 아직 85674명 데이터 남음.
15000명 진행됨. 아직 80674명 데이터 남음.
20000명 진행됨. 아직 75674명 데이터 남음.
25000명 진행됨. 아직 70674명 데이터 남음.
30000명 진행됨. 아직 65674명 데이터 남음.
35000명 진행됨. 아직 60674명 데이터 남음.
40000명 진행됨. 아직 55674명 데이터 남음.
45000명 진행됨. 아직 50674명 데이터 남음.
50000명 진행됨. 아직 45674명 데이터 남음.
55000명 진행됨. 아직 40674명 데이터 남음.
60000명 진행됨. 아직 35674명 데이터 남음.
65000명 진행됨. 아직 30674명 데이터 남음.
70000명 진행됨. 아직 25674명 데이터 남음.
75000명 진행됨. 아직 20674명 데이터 남음.
80000명 진행됨. 아직 15674명 데이터 남음.
85000명 진행됨. 아직 10674명 데이터 남음.
90000명 진행됨. 아직 5674명 데이터 남음.
95000명 진행됨. 아직 674명 데이터 남음.


In [24]:
df_test_model.head()

VisitNumber  -1.0  1.0  0.0  2.0  4.0  5.0  6.0  7.0  8200.0   ...    \
0            1     0    0    0    0    0    0    0    0       0   ...     
1            2     0    0    0    0    0    0    0    0       0   ...     
2            3     0    0    0    0    0    0    0    0       0   ...     
3            4     0    0    0    0    0    0    0    0       0   ...     
4            6     0    0    0    0    0    0    0    0       0   ...     

   8173.0  8175.0  8176.0  1625.0  8180.0  1626.0  8185.0  1627.0  8190.0  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   8191.0  
0       0  
1       0  
2       0  
3       0  
4       0  

[5 rows x 5356 columns]

In [29]:
saveDataFrameToCsv(df_test_model, "df_fl_model_test")

In [20]:
len(np.zeros(len(dd_cols) - 1))

5354

In [22]:
def makeDfWeWanted_local(df, df_train, df_test, group_by_col_name = "DepartmentDescription", is_use_negative_scan_count = False):
    """
        패러미터 정보 : 
        df에는 변환하고자하는 df를 넣어준다.
        df_train에는 train데이터 프레임을 반드시 넣어준다.
        is_use_negative_scan_count는 False가 default이다.
    """
    cols = df.columns
    df_ = df.copy()
    dd_cols = []
    if group_by_col_name == "DepartmentDescription":
        df_[group_by_col_name] = df_[group_by_col_name].apply(lambda a : "Null" if type(a) is float else a)
        dd_cols = [uq if not type(uq) is float else "Null" for uq in df_train[group_by_col_name].unique()]
    else:
        df_[group_by_col_name] = df_[group_by_col_name].apply(lambda a : -1.0 if np.isnan(a) else a)
        train_fl_li = df_train[group_by_col_name].unique()
        test_fl_li = df_test[group_by_col_name].unique()
        dd_cols = [li if not np.isnan(li) else -1.0 for li in list(set(list(test_fl_li) + list(train_fl_li)))]
#         dd_cols = [str(uq) if not np.isnan(uq) else "-1.0" for uq in df_train[group_by_col_name].unique()]
    dd_cols.insert(0, "VisitNumber")
    dd_cols.append("TripType")
    result_df = pd.DataFrame(columns = dd_cols)
    dict_ = dict(zip(dd_cols, np.zeros(len(dd_cols)).astype(int)))
    is_test_df = False
    if "TripType" in cols:
        df_ = pd.DataFrame(df_.groupby(["VisitNumber", "TripType", group_by_col_name]).sum()["ScanCount"]).reset_index()
    else:
        is_test_df = True
        dd_cols = dd_cols[:-1]
        df_ = pd.DataFrame(df_.groupby(["VisitNumber", group_by_col_name]).sum()["ScanCount"]).reset_index()
    return __makeDf_local(df_, is_use_negative_scan_count, dd_cols, is_test_df, group_by_col_name)

def __makeDf(df, is_use_negative_scan_count, dd_cols, is_test_df, group_by_col_name):
    display(Markdown("##### makeDfWeWanted함수는 !! 데이터프레임을 반환하는 \
    함수이므로 변수로 받아주셔야합니다!! 오래 걸리는 작업(맥북 프로 기준 3분)이므로 지금 실수하셨다면 빨리 커널 종료하고 다시 시도해주세요."))
    vn_uq_li = df["VisitNumber"].unique()
    li = []
    scan_count_min = 0
    if is_use_negative_scan_count:
        scan_count_min = -df["ScanCount"].min()
    for i, vn in enumerate(vn_uq_li):
        tmp_df = df[df["VisitNumber"] == vn]
        space = []
        if is_test_df:
            space = (np.zeros(len(dd_cols)).astype(int) + scan_count_min) if not is_test_df else (np.zeros(len(dd_cols)).astype(int) + scan_count_min)
        else:
            space = (np.zeros(71).astype(int) + scan_count_min) if not is_test_df else (np.zeros(70).astype(int) + scan_count_min)
        space[0] = vn
        if not is_test_df:
            tripType = tmp_df["TripType"].unique()[0]
            space[len(space) - 1] = tripType
        for row_nbr in tmp_df.index:
            dd = tmp_df.loc[row_nbr][group_by_col_name]
            scan_cnt = tmp_df.loc[row_nbr]["ScanCount"] + scan_count_min
            idx = dd_cols.index(dd)
            if is_use_negative_scan_count:
                space[idx] = scan_cnt
            else:
                if scan_cnt > 0:
                    space[idx] = scan_cnt
        li.append(space)
        if (i % 5000) == 0:    
            print(str(i) + "명 진행됨. 아직 " + str(len(vn_uq_li) - i) + "명 데이터 남음.")
    return pd.DataFrame(li, columns=dd_cols) 

             precision    recall  f1-score   support      

          3       0.87      0.95      0.91      3643
          4       0.71      0.26      0.38       346
          5       0.72      0.71      0.72      4593
          6       0.71      0.59      0.64      1277
          7       0.69      0.72      0.71      5752
          8       0.39      0.59      0.47     12161
          9       0.62      0.58      0.60      9464
         12       0.56      0.34      0.42       269
         14       1.00      0.75      0.86         4
         15       0.52      0.60      0.56       978
         18       0.60      0.42      0.50       549
         19       0.83      0.25      0.38       375
         20       0.69      0.46      0.55       637
         21       0.59      0.65      0.62       641
         22       0.73      0.43      0.54       928
         23       0.84      0.12      0.20       139
         24       0.59      0.60      0.59      2609
         25       0.60      0.68      0.64      3698
         26       0.69      0.41      0.51       503
         27       0.61      0.61      0.61       785
         28       0.74      0.45      0.56       492
         29       0.85      0.49      0.62       433
         30       0.77      0.32      0.45      1081
         31       0.83      0.65      0.73       594
         32       0.65      0.70      0.67      1984
         33       0.46      0.61      0.52      1315
         34       0.48      0.59      0.53       719
         35       0.47      0.51      0.49      2030
         36       0.44      0.69      0.54      3005
         37       0.50      0.72      0.59      2788
         38       0.55      0.64      0.59      2912
         39       0.41      0.41      0.41      9896
         40       0.52      0.53      0.52      6130
         41       0.77      0.24      0.36       583
         42       0.64      0.30      0.41      1858
         43       0.66      0.14      0.23       872
         44       0.66      0.35      0.46      1187
        999       0.62      0.21      0.32      8444

    avg / total       0.57      0.55      0.54     95674


          precision    recall  f1-score   support

          3       0.80      0.87      0.84      3643
          4       0.07      0.04      0.05       346
          5       0.75      0.81      0.78      4593
          6       0.48      0.55      0.51      1277
          7       0.60      0.66      0.63      5752
          8       0.36      0.52      0.42     12161
          9       0.57      0.45      0.50      9464
         12       0.15      0.21      0.18       269
         14       0.05      0.50      0.09         4
         15       0.37      0.49      0.42       978
         18       0.38      0.53      0.44       549
         19       0.39      0.23      0.29       375
         20       0.47      0.73      0.57       637
         21       0.41      0.76      0.53       641
         22       0.45      0.22      0.30       928
         23       0.27      0.20      0.23       139
         24       0.51      0.64      0.57      2609
         25       0.56      0.74      0.64      3698
         26       0.36      0.51      0.42       503
         27       0.49      0.71      0.58       785
         28       0.38      0.52      0.44       492
         29       0.20      0.09      0.12       433
         30       0.40      0.37      0.38      1081
         31       0.65      0.82      0.72       594
         32       0.59      0.78      0.67      1984
         33       0.39      0.66      0.49      1315
         34       0.39      0.74      0.51       719
         35       0.50      0.54      0.52      2030
         36       0.44      0.65      0.53      3005
         37       0.44      0.67      0.53      2788
         38       0.35      0.56      0.43      2912
         39       0.39      0.21      0.27      9896
         40       0.41      0.30      0.35      6130
         41       0.32      0.07      0.11       583
         42       0.48      0.10      0.16      1858
         43       0.15      0.04      0.06       872
         44       0.40      0.17      0.24      1187
        999       0.09      0.03      0.05      8444

    avg / total       0.44      0.47      0.44     95674